fix: exemption race conditions

TristanHoladay opened the pull request:
## Description
Fixes race conditions with exemptions that result in overwrites of previous exemptions in the Pepr store or mutating then allowing pods that were meant to be exempted from mutation.

## Issue

Fixes #409 

Fixes https://github.com/defenseunicorns/uds-core/issues/314

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [x] Test, docs, adr added or updated as needed
- [ ] [Contributor Guide Steps](https://github.com/defenseunicorns/uds-template-capability/blob/main/CONTRIBUTING.md)(https://github.com/defenseunicorns/uds-template-capability/blob/main/CONTRIBUTING.md#submitting-a-pull-request) followed

TristanHoladay committed:
fix: exemption store overwrites

TristanHoladay committed:
Merge branch 'main' into fix-exemptions

TristanHoladay committed:
refactor to kfc watch

TristanHoladay committed:
pepr format

TristanHoladay committed:
update reconciler tests; added WatchPhase enum to exemptions.

TristanHoladay committed:
Merge branch 'main' into fix-exemptions

cmwylie19 commented on src/pepr/policies/index.ts:
```diff
@@ -1,6 +1,31 @@
+
+    // This will run until the process is terminated or the watch is aborted
+    await watcher.start();
+  }
```
```suggestion
  // only run in admission controller
  if (process.env.PEPR_WATCH_MODE === "false") {
    const watcher = K8s(UDSExemption).Watch(async (exemption, phase) => {
      console.log(`Exemption ${exemption.metadata?.name} is ${phase}`);

      processExemptions(exemption, phase, policyExemptionMap);
    });

    // This will run until the process is terminated or the watch is aborted
    await watcher.start();
  }
```

TristanHoladay commented on src/pepr/operator/controllers/exemptions/exemptions.ts:
```diff
@@ -1,9 +1,14 @@
+import { PolicyMap, StoredMatcher } from "../../../policies";
+import { ExemptionElement, Policy, UDSExemption } from "../../crd";
+
+export enum WatchPhase {
```
Was trying to use the enum from the fluent client but got errors when importing the type. so i just remade it here.

TristanHoladay committed:
Update src/pepr/policies/index.ts

Co-authored-by: Case Wylie <cmwylie19@gmail.com>

mjnagel commented on src/pepr/operator/reconcilers/index.ts:
```diff
@@ -12,7 +12,7 @@ const uidSeen = new Set<string>();
  * @returns true if the CRD is pending or the current generation has been processed
  */
-export function shouldSkip(cr: UDSExemption | UDSPackage) {
+export function shouldSkip(cr: GenericKind) {
```
Should this still be `UDSPackage` instead of `GenericKind`?

TristanHoladay replied:
i went back and forth on that one. I don't think it hurts to set it to `GenericKind` but i'll change it to `UDSPackage`. it'll be more consistent with the other functions that way.

mjnagel reviewed (commented):
One other broad question/thing to validate - would want to make sure that this no longer being queue based doesn't lead to any weird behaviors with rapid create/edit/delete events (today those get processed in order on a queue with `Reconcile`, unsure exactly how that would behave with KFC watch here).

TristanHoladay committed:
fix tests, clean up validator

TristanHoladay commented:
I was thinking the same @mjnagel . so far i haven't seen any issues, but i'll see about creating a test for it.

TristanHoladay committed:
refactor to exemption controller logic.

TristanHoladay committed:
Merge branch 'main' into fix-exemptions

TristanHoladay committed:
format and lint

TristanHoladay committed:
adding logging to exemption controller; podinfo back to back exemption test.

TristanHoladay committed:
add verify for podinfo in exempted-app test

mjnagel commented on src/pepr/operator/controllers/exemptions/exemptions.ts:
```diff
@@ -1,160 +1,114 @@
+    for (const p of policies) {
+      const storedMatchers = map.get(p) ?? [];
+      if (log) {
+        console.log(`Adding to ${p}: ${JSON.stringify([...storedMatchers, matcherToStore])}`);
```
Still working my way through the rest of the PR, but one comment upfront - anywhere we are using `console.log` I think we should really be using `Log.debug` or `Log.info` depending on the specific case. This will ensure we have consistent logging and are using the same log method as all the other parts of the operator + admission pieces.

I'm also on the fence on the conditional for logging here. I think I see how it is used (only logging adds, not modifies?), but would prefer that we err on the side of more logging rather than less.

TristanHoladay replied:
good point about the logger. I forgot about `Log` and just defaulted to `console`. I added the boolean so that we don't log adding to the temporary map. the updates will still get logged, just not from the `addToMap`.

mjnagel replied:
Ah okay, tracking. Sounds good.

mjnagel commented on src/pepr/operator/crd/validators/exempt-validator.ts:
```diff
@@ -30,11 +30,6 @@ export async function exemptValidator(req: PeprValidateRequest<UDSExemption>) {
     }
   }
 
-  // Validate there's at least 1 exemption element
-  if (exemptions.length === 0) {
-    return req.Deny("Invalid number of exemptions: must have at least 1");
-  }
-
```
Is there a specific reason this validation was removed? I don't think we strictly need to block empty exemptions (although it could be confusing), just wondering if I'm missing something here that necessitated this change?

Worth testing this but I think we can actually add `minItems: 1` into the exemption source (lined up with the `type: "array"`) and that should enforce at the schema level rather than here.

mjnagel commented on src/test/tasks.yaml:
```diff
@@ -49,5 +49,15 @@ tasks:
+            kind: Pod
+            name: ${PODINFO_NAME}
+            namespace: podinfo
+
```
This failed for me locally on the first run:
```
Error creating: admission webhook "pepr-uds-core.pepr.dev" denied the request: Host ports are not allowed.; Unauthorized container securityContext. Containers must not run as root. Authorized: [runAsNonRoot = false | runAsUser > 0] Found: {"name":"podinfo","ctx":{"capabilities":{"drop":["ALL"]},"runAsUser":0,"runAsGroup":0,"runAsNonRoot":false}}
```

I see all 5 exemption resources created, and actually see `Adding to RequireNonRootUser:` with podinfo in the logs too 🤔. I'm going to re-run a few times and look through the logic but it worries me a bit that this could fail 😬.

mjnagel replied:
Tentatively going to resolve this, I haven't been able to reproduce this issue and have made a few refactors locally so am hoping this is a non-issue at this point. Something to keep an eye on in CI/local dev though.

TristanHoladay committed:
replacing console.log with Log

mjnagel commented on src/test/tasks.yaml:
```diff
@@ -49,5 +49,15 @@ tasks:
+          cluster:
+            kind: Pod
+            name: ${PODINFO_NAME}
+            namespace: podinfo
```
```suggestion
      - description: Check for podinfo to come up
        wait:
          cluster:
            kind: Pod
            name: app.kubernetes.io/name=podinfo
            namespace: podinfo
```
nit: this supports labels so we should be able to skip the "get pod name" part.

mjnagel commented on src/test/tasks.yaml:
```diff
@@ -49,5 +49,15 @@ tasks:
+            namespace: podinfo
+
       - description: Remove the test resources
         cmd: "uds zarf package remove build/zarf-package-uds-core-test-apps-*.zst --confirm"
```
```suggestion
        cmd: "uds zarf package remove build/zarf-package-uds-core-test-apps-*.zst --confirm --no-progress"
```
would be a nice addition while we're modifying this 😄

TristanHoladay committed:
adding logging to isExempt() check

mjnagel committed:
fix: tasks, schema [ci skip]

mjnagel committed:
wip: fix update issue

mjnagel committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
some more refactors

mjnagel committed:
use ramda for list equality/union

mjnagel committed:
fix test

TristanHoladay committed:
Merge branch 'main' into fix-exemptions

rjferguson21 committed:
Exemption store (#421)

## Description
Refactor some exemption logic

## Related Issue

Fixes #
<!-- or -->
Relates to #

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [ ] Test, docs, adr added or updated as needed
- [ ] [Contributor Guide
Steps](https://github.com/defenseunicorns/uds-template-capability/blob/main/CONTRIBUTING.md)(https://github.com/defenseunicorns/uds-template-capability/blob/main/CONTRIBUTING.md#submitting-a-pull-request)
followed

---------

Co-authored-by: Micah Nagel <micah.nagel@defenseunicorns.com>

mjnagel committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
rebase

mjnagel committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
Merge branch 'main' into fix-exemptions

rjferguson21 committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
chore: move CRD apply to admission controller

mjnagel committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
chore: switch crd register to be more visible/timed well

mjnagel committed:
chore: timing debug

mjnagel committed:
pepr format

mjnagel committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
wip: iife CRD registration timing

mjnagel committed:
Merge branch 'main' into fix-exemptions

mjnagel committed:
error handling

mjnagel committed:
chore: peprstore cleanup

mjnagel committed:
Merge branch 'main' into fix-exemptions
